File: python/dolfin/function/constant.py
# -*- coding: utf-8 -*-
# Copyright (C) 2019 Michal Habera and Chris Richardson
#
# This file is part of DOLFIN (https://www.fenicsproject.org)
#
# SPDX-License-Identifier:    LGPL-3.0-or-later

from typing import Sequence, Union

import numpy

import dolfin.cpp
import ufl


class Constant(ufl.Constant):
    def __init__(self, domain, value: Union[numpy.ndarray, Sequence, float]):
        """A constant with respect to a domain.

        Parameters
        ----------
        domain : DOLFIN or UFL mesh
        value
            Value of the constant.
        """
        np_value = numpy.asarray(value)
        super().__init__(domain, np_value.shape)
        self._cpp_object = dolfin.cpp.function.Constant(np_value.shape, np_value.flatten())

    @property
    def value(self):
        """Returns value of the constant."""
        return self._cpp_object.value()

    @value.setter
    def value(self, v):
        numpy.copyto(self._cpp_object.value(), numpy.asarray(v))

File: cpp/dolfin/function/Constant.cpp
// Copyright (C) 2019 Chris Richardson abd Michal Habera
//
// This file is part of DOLFIN (https://www.fenicsproject.org)
//
// SPDX-License-Identifier:    LGPL-3.0-or-later

#include "Constant.h"
#include <Eigen/Dense>
#include <vector>

using namespace dolfin;
using namespace dolfin::function;

//-----------------------------------------------------------------------------
Constant::Constant(PetscScalar value) : value({value})
{
  // Do nothing
}
//-----------------------------------------------------------------------------
Constant::Constant(std::vector<PetscScalar> value, std::vector<int> shape)
    : value(value), shape(shape)
{
  // Do nothing
}
//-----------------------------------------------------------------------------
Constant::Constant(
    const Eigen::Ref<Eigen::Array<PetscScalar, Eigen::Dynamic, Eigen::Dynamic,
                                  Eigen::RowMajor>>
        c)
    : value(c.rows() * c.cols()), shape({(int)c.rows(), (int)c.cols()})
{
  // Copy data from Eigen::Array to flattened vector
  for (int i = 0; i < c.rows(); ++i)
    for (int j = 0; j < c.cols(); ++j)
      value[i * c.cols() + j] = c(i, j);
}
//-----------------------------------------------------------------------------
